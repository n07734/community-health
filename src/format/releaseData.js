import { propOr } from 'ramda'
import { diff, lt } from 'semver'

const datePropSort = (a, b) => {
    const getDate = propOr('', 'date')
    return new Date(getDate(a)).getTime() - new Date(getDate(b)).getTime()
}

const getTag = propOr('', 'description')

const semVerSort = (a, b) => {
    let sortResult
    try {
        sortResult = lt(getTag(a), getTag(b))
            ? -1
            : 1
    } catch (error) {
        sortResult = 0
    }

    return sortResult
}

const formatReleaseData = (results = []) => {
    const semVerSortedResults = results
        .filter(x => /^v\d+\.\d+\.\d+$/.test(getTag(x)))
        .sort(semVerSort)

    const formattedReleases = semVerSortedResults
        .map((release = {}, i) => {
            const currentTag = getTag(release)
            let releaseType = ''

            try {
                const prevTag = i > 0
                    ? getTag(semVerSortedResults[i - 1])
                    : 'v0.0.0'

                const diffType = diff(prevTag, currentTag) || ''

                 
                releaseType = ({
                    'minor': 'MINOR',
                    'major': 'MAJOR',
                })[diffType] || 'PATCH'

            } catch (error) {
                // eslint-disable-next-line no-unused-vars
                releaseType = 'PATCH'
            }

            return {
                ...release,
                releaseType: release.releaseType || 'MAJOR',
            }
        })

    const finalFilter = formattedReleases.length > 70
        ? (x = {}) => x.releaseType !== 'PATCH'
        : () => true

    const dateSortedReleases = formattedReleases
        .filter(finalFilter)
        .sort(datePropSort)

    return dateSortedReleases
}

const formatMarkers = ({ usersInfo = {}, events = [] } = {}) => {
    const teamMarkers = []
    Object.values(usersInfo)
        .forEach(({
            name,
            userId,
            dates = [],
        } = {}) => {
            dates
                .forEach((dateInfo = {}) => {
                    const {
                        startDate,
                        endDate,
                    } = dateInfo
                    if (startDate) {
                        teamMarkers.push({
                            date: startDate,
                            description: name || userId,
                            releaseType: 'MINOR',
                        })
                    }
                    if (endDate) {
                        teamMarkers.push({
                            date: endDate,
                            description: name || userId,
                            releaseType: 'PATCH',
                        })
                    }
                })
        })

    const eventMarkers = events
        .map(({date, name} = {}) => ({
            date,
            description: name,
            releaseType: 'MAJOR',
        }))
    const dateSort = ({ date: a }, { date: b }) => new Date(a).getTime() - new Date(b).getTime()
    const sortedMarkers = [...teamMarkers, ...eventMarkers].sort(dateSort)

    return sortedMarkers
}

export {
    formatMarkers,
    formatReleaseData,
}
