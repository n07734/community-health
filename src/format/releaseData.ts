import { diff, lt } from 'semver'
import { UserDate, UserInfo, UsersInfo } from '../types/State'
import { ObjStrings } from '../types/Components'
import { EventInfo } from '../types/FormattedData'

const datePropSort = (aObj: ObjStrings, bObj: ObjStrings) => {
    const a = aObj?.date || ''
    const b = bObj?.date || ''
    return new Date(a).getTime() - new Date(b).getTime()
}

const getTag = (arg: ObjStrings): string => {
    const value = arg?.description || ''
    return value as string
}

const semVerSort = (a: ObjStrings, b: ObjStrings) => {
    let sortResult
    try {
        sortResult = lt(getTag(a), getTag(b))
            ? -1
            : 1
    } catch (error) {
        sortResult = 0
    }

    return sortResult
}

type ReleaseResult = {
    description: string
    releaseType: string
}
const formatReleaseData = (results: ReleaseResult[] = []) => {
    const semVerSortedResults = results
        .filter(x => /^v\d+\.\d+\.\d+$/.test(getTag(x)))
        .sort(semVerSort)

    const formattedReleases = semVerSortedResults
        .map((release:ReleaseResult, i) => {
            const currentTag = getTag(release)
            let releaseType = ''

            try {
                const prevTag = i > 0
                    ? getTag(semVerSortedResults[i - 1])
                    : 'v0.0.0'

                const diffType = diff(prevTag, currentTag) || 'EMPTY'

                releaseType = ({
                    'minor': 'MINOR',
                    'major': 'MAJOR',
                    'patch': 'PATCH',
                    'premajor': 'PATCH',
                    'preminor': 'PATCH',
                    'prepatch': 'PATCH',
                    'prerelease': 'PATCH',
                    'EMPTY': 'PATCH',
                })[diffType] || 'PATCH'

            } catch (error) {
                // eslint-disable-next-line no-unused-vars
                releaseType = 'PATCH'
            }

            return {
                ...release,
                releaseType: releaseType || 'MAJOR',
            }
        })

    const finalFilter = formattedReleases.length > 70
        ? (release: any = {}) => release.releaseType !== 'PATCH'
        : () => true

    const dateSortedReleases = formattedReleases
        .filter(finalFilter)
        .sort((a:any,b:any) => datePropSort(a,b))

    return dateSortedReleases
}


const formatMarkers = ({
    usersInfo = {},
    events = []
}: {
    usersInfo?: UsersInfo
    events: EventInfo[]
}) => {
    const teamMarkers: any[] = []
    Object.values(usersInfo)
        .forEach(({
            name = '',
            userId = '',
            dates = [],
        }: UserInfo) => {
            dates
                .forEach((dateInfo:UserDate) => {
                    const {
                        startDate,
                        endDate,
                    } = dateInfo
                    if (startDate) {
                        teamMarkers.push({
                            date: startDate,
                            description: name || userId,
                            releaseType: 'MINOR',
                        })
                    }
                    if (endDate) {
                        teamMarkers.push({
                            date: endDate,
                            description: name || userId,
                            releaseType: 'PATCH',
                        })
                    }
                })
        })

    const eventMarkers = events
        .map(({date, name}: ObjStrings = {}):EventInfo => ({
            date,
            description: name,
            releaseType: 'MAJOR',
        }))
    const dateSort = (aObj: ObjStrings, bObj: ObjStrings) => {
        const a = aObj?.date || ''
        const b = bObj?.date || ''
        return new Date(a).getTime() - new Date(b).getTime()
    }
    const sortedMarkers = [...teamMarkers, ...eventMarkers].sort(dateSort)

    return sortedMarkers
}

export {
    formatMarkers,
    formatReleaseData,
}
