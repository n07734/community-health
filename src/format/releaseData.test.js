import releaseData from './releaseData'

describe('releaseData:', () => {
    it('correct versioning and date sorting', () => {
        const data = [
            { description: 'v0.0.1', date: '2020-10-30' },
            { description: 'v0.2.0', date: '2020-10-29' },
            { description: 'v0.2.1', date: '2020-10-28' },
            { description: 'v5.3.0', date: '2020-10-20' },
            { description: 'v1.3.0', date: '2020-10-19' },
            { description: 'v1.3.1', date: '2020-10-18' },
            { description: 'v2.0.0', date: '2020-10-15' },
            { description: 'v3.0.0', date: '2020-10-10' },
        ]

        const result = releaseData(data)

        expect(result).toMatchSnapshot()
    })

    it('ignore invalid tags', () => {
        const data = [
            { description: 'v0.0.1', date: '2020-10-30' },
            { description: 'v0.2.UU', date: '2020-10-29' },
            { description: 'v0.3.0', date: '2020-10-20' },
        ]

        const result = releaseData(data)

        expect(result).toMatchSnapshot()
    })
})