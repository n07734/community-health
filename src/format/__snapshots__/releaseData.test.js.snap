// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`releaseData: correct versioning and date sorting 1`] = `
[
  {
    'date': '2020-10-10',
    'description': 'v3.0.0',
    'releaseType': 'MAJOR',
  },
  {
    'date': '2020-10-15',
    'description': 'v2.0.0',
    'releaseType': 'MAJOR',
  },
  {
    'date': '2020-10-18',
    'description': 'v1.3.1',
    'releaseType': 'PATCH',
  },
  {
    'date': '2020-10-19',
    'description': 'v1.3.0',
    'releaseType': 'MAJOR',
  },
  {
    'date': '2020-10-20',
    'description': 'v5.3.0',
    'releaseType': 'MAJOR',
  },
  {
    'date': '2020-10-28',
    'description': 'v0.2.1',
    'releaseType': 'PATCH',
  },
  {
    'date': '2020-10-29',
    'description': 'v0.2.0',
    'releaseType': 'MINOR',
  },
  {
    'date': '2020-10-30',
    'description': 'v0.0.1',
    'releaseType': 'PATCH',
  },
]
`;

exports[`releaseData: ignore invalid tags 1`] = `
[
  {
    'date': '2020-10-20',
    'description': 'v0.3.0',
    'releaseType': 'MINOR',
  },
  {
    'date': '2020-10-30',
    'description': 'v0.0.1',
    'releaseType': 'PATCH',
  },
]
`;
